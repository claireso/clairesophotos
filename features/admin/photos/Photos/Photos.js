import { Fragment, useEffect, useCallback } from 'react'
import { useRouter } from 'next/router'

import PhotosReducer, { ACTION_TYPES } from '../reducer'

import Loader from '@components/Loader'
import { IconPlus } from '@components/Icons'
import { List, ListHeader } from '@components/List'
import { PrimaryButton } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Pager from '@components/Pager'

import Photo from './components/Photo'

const Photos = () => {
  const [
    { items: photos, pager, ...state },
    { loadResources }
  ] = PhotosReducer.usePhotosReducer()
  const isLoading = ['idle', 'loading'].includes(state.status)

  const router = useRouter()
  const {
    query: { page },
    pathname
  } = router

  useEffect(() => {
    loadResources(page)
  }, [page, loadResources])

  const navigate = useCallback(
    (params) => {
      const query = {}
      if (page) query.page = page
      router.push({ pathname: pathname, query: { ...query, ...params } })
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [page]
  )

  const onPageChange = (page) => navigate({ page })

  const onCreate = useCallback(
    (ev) => {
      ev.preventDefault()
      navigate({ action: ACTION_TYPES.CREATE })
    },
    [navigate]
  )

  const onDelete = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.DELETE,
        id: id
      })
    },
    [navigate]
  )

  const onEdit = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.EDIT,
        id: id
      })
    },
    [navigate]
  )

  return (
    <Fragment>
      <ListHeader>
        <Heading1 data-testid="list-heading">
          Your photos {pager && <span>({pager.count})</span>}
        </Heading1>
        <PrimaryButton data-testid="button-create" onClick={onCreate}>
          Add a new photo
          <IconPlus />
        </PrimaryButton>
      </ListHeader>

      {isLoading ? (
        <Loader />
      ) : (
        <Fragment>
          <List>
            {photos.map((photo, index) => (
              <Photo
                key={index}
                {...photo}
                onDelete={onDelete}
                onEdit={onEdit}
              />
            ))}
          </List>
          <Pager {...pager} navigate={onPageChange} />
        </Fragment>
      )}
    </Fragment>
  )
}

export default Photos
