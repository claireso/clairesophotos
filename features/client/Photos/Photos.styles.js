import styled from 'styled-components'

export const PagerWrapper = styled.div`
  grid-column: 1 / -1;
  margin: 4.5rem 0;

  @media (min-width: 800px) {
    margin: 8.5rem 0;
  }
`
