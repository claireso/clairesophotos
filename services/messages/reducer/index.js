import { useReducer, useCallback } from 'react'
import emitter from '@services/emitter'

const INITIAL_STATE = []

const ACTION_ADD_MESSAGE = 'messages/add'
const ACTION_CLOSE_MESSAGE = 'messages/remove'
const ACTION_CLOSE_ALL = 'messages/close_all'

function reducer(state, { type, ...message }) {
  switch (type) {
    case ACTION_ADD_MESSAGE: {
      let index = state.length

      if (message.key) {
        const newIndex = state.findIndex((m) => m.key === message.key)

        if (newIndex > -1) {
          index = newIndex
        }
      }

      return [...state.slice(0, index), message, ...state.slice(index + 1)]
    }

    case ACTION_CLOSE_MESSAGE: {
      const index = message.index

      return [...state.slice(0, index), ...state.slice(index + 1)]
    }

    case ACTION_CLOSE_ALL: {
      return []
    }

    default:
      return [...state]
  }
}

export const useMessagesReducer = () => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE)

  const displayMessage = useCallback(
    (message) => dispatch({ type: ACTION_ADD_MESSAGE, ...message }),
    [dispatch]
  )
  const closeMessage = useCallback(
    (index) => dispatch({ type: ACTION_CLOSE_MESSAGE, index }),
    [dispatch]
  )
  const closeAllMessages = useCallback(
    () => dispatch({ type: ACTION_CLOSE_ALL }),
    [dispatch]
  )

  return [state, { displayMessage, closeMessage, closeAllMessages }]
}

const displayMessage = (status) => (message) => {
  emitter.emit('feat/messages/add', {
    status,
    ...message
  })
}

export const displaySuccessMessage = displayMessage('success')
export const displayErrorMessage = displayMessage('error')
