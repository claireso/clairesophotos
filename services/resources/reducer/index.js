import React, { useReducer, useContext, useCallback } from 'react'
import PropTypes from 'prop-types'

const STATUS_IDLE = 'idle'
const STATUS_PENDING = 'pending'
const STATUS_LOADING = 'loading'
const STATUS_ERROR = 'error'
const STATUS_UPDATE_DATA_SUCCESS = 'resources/update_success'

const ACTION_UPDATE_STATUS = 'resources/update_status'
const ACTION_LOAD_RESOURCES = 'resources/get_all'
const ACTION_LOAD_RESOURCE = 'resources/get'
const ACTION_ADD_RESOURCE = 'resources/add'
const ACTION_DELETE_RESOURCE = 'resources/delete'
const ACTION_EDIT_RESOURCE = 'resources/edit'

const INITIAL_STATE = {
  status: STATUS_IDLE,
  items: [],
  pager: {}
}

export const reducer = (state, action) => {
  switch (action.type) {
    case ACTION_UPDATE_STATUS:
      return {
        ...state,
        status: action.status
      }

    case ACTION_LOAD_RESOURCES:
      return {
        ...state,
        ...action.response,
        status: STATUS_UPDATE_DATA_SUCCESS
      }

    case ACTION_LOAD_RESOURCE:
      return {
        ...state,
        single: action.response
      }

    case ACTION_ADD_RESOURCE:
      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [action.response, ...state.items],
        pager: {
          ...state.pager,
          count: state.pager.count + 1
        }
      }

    case ACTION_EDIT_RESOURCE: {
      const index = state.items.findIndex((p) => p.id == action.response.id)

      if (index < 0) return { ...state }

      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [
          ...state.items.slice(0, index),
          action.response,
          ...state.items.slice(index + 1)
        ]
      }
    }

    case ACTION_DELETE_RESOURCE: {
      const index = state.items.findIndex((p) => p.id === action.id)

      if (index < 0) return { ...state }

      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [
          ...state.items.slice(0, index),
          ...state.items.slice(index + 1)
        ],
        pager: {
          ...state.pager,
          count: state.pager.count - 1
        }
      }
    }
    default:
      return { ...state }
  }
}

export const createResourceManager = (options = {}) => {
  const ResourcesStateContext = React.createContext()
  const ResourcesDispatchContext = React.createContext()

  const ResourcesProvider = ({ children, value }) => {
    const [state, dispatch] = useReducer(reducer, value)

    return (
      <ResourcesStateContext.Provider value={state}>
        <ResourcesDispatchContext.Provider value={dispatch}>
          {children}
        </ResourcesDispatchContext.Provider>
      </ResourcesStateContext.Provider>
    )
  }

  ResourcesProvider.propTypes = {
    children: PropTypes.any,
    value: PropTypes.any
  }

  const useResourcesState = () => {
    const context = useContext(ResourcesStateContext)

    if (context === undefined) {
      throw new Error(
        'useResourcesState must be used within a ResourcesProvider'
      )
    }

    return context
  }

  const useResourcesDispatch = () => {
    const context = useContext(ResourcesDispatchContext)

    if (context === undefined) {
      throw new Error(
        'useResourcesDispatch must be used within a ResourcesProvider'
      )
    }

    return context
  }

  const getAction = (name) => {
    const actionsOptions = options.actions || {}
    const action = actionsOptions[name]

    if (!action || !action.action) {
      throw new Error(
        `ResourceManager: missing options for the action '${name}'`
      )
    }

    return action
  }

  const onSuccess = (action) => {
    if (action.onSuccess) {
      const message = action.onSuccess()
      if (message) {
        message
      }
    }
  }

  const onError = (action, err) => {
    if (err?.response?.status === 401) return

    if (action.onError) {
      const message = action.onError()
      if (message) {
        message
      }
    }
  }

  const onEnd = (action) => {
    action.onEnd && action.onEnd()
  }

  const useResourcesReducer = () => {
    const state = useResourcesState()
    const dispatch = useResourcesDispatch()

    const loadResources = useCallback(
      async (page = 1) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_LOADING })

        const action = getAction('loadResources')

        try {
          let response = await action.action(page)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_LOAD_RESOURCES, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const loadResource = useCallback(
      async (id) => {
        const action = getAction('loadResource')

        try {
          let response = await action.action(id)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_LOAD_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const createResource = useCallback(
      async (data) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('createResource')

        try {
          let response = await action.action(data)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_ADD_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const editResource = useCallback(
      async (id, data) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('editResource')

        try {
          let response = await action.action(id, data)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_EDIT_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const deleteResource = useCallback(
      async (id) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('deleteResource')

        try {
          await action.action(id)
          dispatch({ type: ACTION_DELETE_RESOURCE, id })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })
          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const test = () => {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })
    }

    return [
      state,
      {
        loadResources,
        loadResource,
        createResource,
        editResource,
        deleteResource,
        test
      }
    ]
  }

  return {
    INITIAL_STATE: { ...INITIAL_STATE },
    Provider: ResourcesProvider,
    useReducer: useResourcesReducer
  }
}
