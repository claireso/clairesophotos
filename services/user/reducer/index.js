import React, { useReducer, useContext, useCallback } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import { loginUser, getMe, logout } from '@services/api'

const STATUS_IDLE = 'idle'
const STATUS_PENDING = 'pending'

const STATUS_AUTH_SUCCESS = 'auth_success'
const STATUS_AUTH_ERROR = 'auth_error'

const ACTION_UPDATE_STATUS = 'user/update_status'
const ACTION_SET_USER = 'user/set'

export const INITIAL_STATE = {
  status: STATUS_IDLE
}

const UserStateContext = React.createContext()
const UserDispatchContext = React.createContext()

function reducer(state, action) {
  switch (action.type) {
    case ACTION_UPDATE_STATUS:
      return {
        ...state,
        status: action.status
      }
    case ACTION_SET_USER: {
      const user = action.user

      return {
        ...state,
        status: user ? STATUS_AUTH_SUCCESS : STATUS_IDLE,
        user: user
      }
    }
    default:
      return { ...state }
  }
}

export const UserProvider = ({ children, value }) => {
  const [state, dispatch] = useReducer(reducer, value)

  return (
    <UserStateContext.Provider value={state}>
      <UserDispatchContext.Provider value={dispatch}>
        {children}
      </UserDispatchContext.Provider>
    </UserStateContext.Provider>
  )
}

UserProvider.propTypes = {
  children: PropTypes.any,
  value: PropTypes.object
}

const useUserState = () => {
  const context = useContext(UserStateContext)

  if (context === undefined) {
    throw new Error('useUserState must be used within a UserProvider')
  }

  return context
}

const useUserDispatch = () => {
  const context = useContext(UserDispatchContext)

  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider')
  }

  return context
}

export const useUserReducer = () => {
  const state = useUserState()
  const dispatch = useUserDispatch()

  const actionLogin = useCallback(
    async (data) => {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

      try {
        await loginUser(data).ready
        const user = await getMe().ready

        dispatch({ type: ACTION_SET_USER, user })

        Router.push('/admin/photos')
      } catch (err) {
        const status = err?.response?.status

        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_AUTH_ERROR })

        if ([401, 422].includes(status)) {
          alert('Bad username/password. Please retry')
        }
      }
    },
    [dispatch]
  )

  const actionLogout = useCallback(async () => {
    try {
      await logout().ready
      dispatch({ type: ACTION_SET_USER, undefined })
    } finally {
      Router.push('/admin/login')
    }
  }, [dispatch])

  const actionGetMe = useCallback(async () => {
    dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

    try {
      const user = await getMe().ready

      dispatch({ type: ACTION_SET_USER, user })
    } catch {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_AUTH_ERROR })
    }
  }, [dispatch])

  return [
    state,
    { login: actionLogin, logout: actionLogout, getMe: actionGetMe }
  ]
}
