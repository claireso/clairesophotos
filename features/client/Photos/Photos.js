import { Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import Pager from '@components/Pager'

import Photo from './components/Photo'

import * as S from './Photos.styles'

const Photos = (props = {}) => {
  const navigate = useCallback((page) => {
    Router.push({ pathname: '/', query: { page } })
    window.scrollTo(0, 0)
  }, [])

  return (
    <Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} row={index + 1} />
      ))}
      <S.PagerWrapper>
        <Pager {...props.pager} navigate={navigate} />
      </S.PagerWrapper>
    </Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
