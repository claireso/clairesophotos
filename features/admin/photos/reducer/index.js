import Router from 'next/router'

import { createResourceManager } from '@services/resources/reducer'
import {
  displaySuccessMessage,
  displayErrorMessage
} from '@services/messages/reducer'
import {
  getPhotos,
  createPhoto,
  deletePhoto,
  editPhoto,
  getPhoto
} from '@services/api'

const PhotosResourceManager = createResourceManager({
  actions: {
    loadResources: {
      action: getPhotos
    },
    loadResource: {
      action: getPhoto
    },
    createResource: {
      action: createPhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been created successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the creation. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    },
    editResource: {
      action: editPhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been updated successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the update. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    },
    deleteResource: {
      action: deletePhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been deleted successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the deletion. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    }
  }
})

function navigateToList({ root = false } = {}) {
  const _query = {}

  if (!root && Router.router.query.page) {
    _query['page'] = Router.router.query.page
  }

  Router.push({ pathname: '/admin/photos', query: _query })

  window.scrollTo(0, 0)
}

export const ACTION_TYPES = {
  CREATE: 'create_photo',
  EDIT: 'edit_photo',
  DELETE: 'delete_photo'
}

export const {
  Provider: PhotosProvider,
  useReducer: usePhotosReducer,
  INITIAL_STATE
} = PhotosResourceManager
