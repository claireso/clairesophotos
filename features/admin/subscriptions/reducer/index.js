import Router from 'next/router'

import { createResourceManager } from '@services/resources/reducer'
import {
  displaySuccessMessage,
  displayErrorMessage
} from '@services/messages/reducer'
import { getSubscriptions, deleteSubscription } from '@services/api'

const SubscriptionsResourceManager = createResourceManager({
  actions: {
    loadResources: {
      action: getSubscriptions
    },
    deleteResource: {
      action: deleteSubscription,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your subscription has been deleted successfully',
          key: 'CRUD_SUBSCRIPTION'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the deletion. Please retry',
          key: 'CRUD_SUBSCRIPTION'
        })
        navigateToList()
      }
    }
  }
})

function navigateToList({ root = false } = {}) {
  const _query = {}

  if (!root && Router.router.query.page) {
    _query['page'] = Router.router.query.page
  }

  Router.push({ pathname: '/admin/subscriptions', query: _query })

  window.scrollTo(0, 0)
}

export const ACTION_TYPES = {
  DELETE: 'delete_subscription'
}

export const {
  Provider: SubscriptionsProvider,
  useReducer: useSubscriptionsReducer,
  INITIAL_STATE
} = SubscriptionsResourceManager
