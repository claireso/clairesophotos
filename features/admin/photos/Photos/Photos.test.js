import { render, waitFor, fireEvent } from '@testing-library/react'

import withTestRouter from '@utils/hoc/withTestRouter'

import { PhotosProvider, INITIAL_STATE } from '../reducer'
import Photos from './Photos'

describe('List Photos', () => {
  const renderComponent = (state = INITIAL_STATE, routerProps = {}) => {
    return render(
      withTestRouter(
        <PhotosProvider value={state}>
          <Photos />
        </PhotosProvider>,
        {
          pathname: '/admin/photos',
          ...routerProps
        }
      )
    )
  }

  beforeEach(() => {
    fetch.resetMocks()
    fetch.mockResponseOnce(JSON.stringify(global.__PHOTOS__))
  })

  test('should render loading view [status - idle]', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })

  test('should render loading view [status - loading]', () => {
    const state = {
      ...INITIAL_STATE,
      status: 'loading'
    }

    const { container } = renderComponent(state)

    expect(container).toMatchSnapshot()
  })

  test('should load and render photos', async () => {
    const { container, getByText } = renderComponent()

    await waitFor(() => {
      expect(getByText('Février 2019')).toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  test('should call edit / delete / create photo actions', async () => {
    const push = jest.fn()

    const { getByText, getAllByTitle } = renderComponent(undefined, { push })

    await waitFor(() => {
      expect(getByText('Février 2019')).toBeInTheDocument()
    })

    fireEvent.click(getByText('Add a new photo'))

    expect(push).toHaveBeenNthCalledWith(1, {
      pathname: '/admin/photos',
      query: {
        action: 'create_photo'
      }
    })

    fireEvent.click(getAllByTitle('Edit')[0])

    expect(push).toHaveBeenNthCalledWith(2, {
      pathname: '/admin/photos',
      query: {
        action: 'edit_photo',
        id: 199
      }
    })

    fireEvent.click(getAllByTitle('Delete')[0])

    expect(push).toHaveBeenNthCalledWith(3, {
      pathname: '/admin/photos',
      query: {
        action: 'delete_photo',
        id: 199
      }
    })
  })

  test('should navigate from pager', async () => {
    const push = jest.fn()

    const { getByTitle, getByText } = renderComponent(undefined, { push })

    await waitFor(() => {
      expect(getByText('Février 2019')).toBeInTheDocument()
    })

    fireEvent.click(getByTitle(/Next page/i))

    expect(push).toHaveBeenCalledWith({
      pathname: '/admin/photos',
      query: {
        page: 2
      }
    })
  })
})
