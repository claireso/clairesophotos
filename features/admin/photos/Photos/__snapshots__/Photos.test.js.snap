// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Photos should render loading view [status - idle] 1`] = `
.c4 {
  margin: 2rem 0;
  -webkit-animation: jBcSpD 800ms ease-in-out infinite alternate;
  animation: jBcSpD 800ms ease-in-out infinite alternate;
}

.c4:after {
  background: var(--primary-normal);
  border-radius: 50%;
  content: '';
  display: block;
  height: 1rem;
  width: 1rem;
  margin: 0 auto;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 2rem;
}

.c0 > h1 {
  line-height: 1.835;
  margin: 0;
}

.c0 > h1 > span {
  color: var(--gray-darker);
  font-size: 2rem;
  font-weight: normal;
  line-height: 1;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  background: var(--primary-normal);
  color: white;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c3 + .c2 {
  margin: 0 0 0 0.8rem;
}

.c3:hover {
  background: var(--primary-lighter);
}

.c1 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
  font-weight: 500;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Your photos 
      <span>
        (
        )
      </span>
    </h1>
    <button
      class="Buttonsstyles__DefaultButton-jpizpn-0 Buttonsstyles__StyledButton-jpizpn-1 c2 c3"
    >
      Add a new photo
      <svg
        height="14"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.984 12.984h-6v6h-1.97v-6h-6v-1.97h6v-6h1.97v6h6v1.97z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c4"
  />
</div>
`;

exports[`List Photos should render loading view [status - loading] 1`] = `
.c4 {
  margin: 2rem 0;
  -webkit-animation: jBcSpD 800ms ease-in-out infinite alternate;
  animation: jBcSpD 800ms ease-in-out infinite alternate;
}

.c4:after {
  background: var(--primary-normal);
  border-radius: 50%;
  content: '';
  display: block;
  height: 1rem;
  width: 1rem;
  margin: 0 auto;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 2rem;
}

.c0 > h1 {
  line-height: 1.835;
  margin: 0;
}

.c0 > h1 > span {
  color: var(--gray-darker);
  font-size: 2rem;
  font-weight: normal;
  line-height: 1;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  background: var(--primary-normal);
  color: white;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c3 + .c2 {
  margin: 0 0 0 0.8rem;
}

.c3:hover {
  background: var(--primary-lighter);
}

.c1 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
  font-weight: 500;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Your photos 
      <span>
        (
        )
      </span>
    </h1>
    <button
      class="Buttonsstyles__DefaultButton-jpizpn-0 Buttonsstyles__StyledButton-jpizpn-1 c2 c3"
    >
      Add a new photo
      <svg
        height="14"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.984 12.984h-6v6h-1.97v-6h-6v-1.97h6v-6h1.97v6h6v1.97z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c4"
  />
</div>
`;
